odl/trafos/convolution.py

```diff
@@ -0,0 +1,804 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Operators based on (integral) transformations."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+import scipy.signal as signal
+
+from odl.discr.discr_ops import Resampling
+from odl.discr.lp_discr import (
+    DiscreteLp, DiscreteLpVector, uniform_discr, uniform_discr_fromdiscr)
+from odl.operator.operator import Operator
+from odl.space.ntuples import Ntuples
+from odl.space.cu_ntuples import CudaNtuples
+from odl.set.sets import ComplexNumbers
+from odl.trafos.fourier import FourierTransform
+from odl.util.normalize import normalized_scalar_param_list, safe_int_conv
+
+
+__all__ = ('Convolution', 'FourierSpaceConvolution', 'RealSpaceConvolution')
+
+
+_REAL_CONV_SUPPORTED_IMPL = ('scipy_convolve',)
+_FOURIER_CONV_SUPPORTED_IMPL = ('default', 'pyfftw')
+_FOURIER_CONV_SUPPORTED_KER_MODES = ('real', 'fourier')
+
+
+class Convolution(Operator):
+
+    """Discretization of the convolution integral as an operator.
+
+    This class provides common attributes and methods for subclasses
+    implementing different variants.
+    """
+
+    def kernel(self):
+        """Return the kernel of this convolution operator."""
+        raise NotImplementedError('abstract method')
+
+
+class FourierSpaceConvolution(Convolution):
+
+    """Convolution with a kernel, implemented in Fourier space.
+
+    This operator implements a discrete approximation to the continuous
+    convolution with a fixed kernel. It supports real-space and
+    Fourier based back-ends.
+    """
+
+    def __init__(self, domain, kernel, **kwargs):
```

Comment:
One way of doing this would be to take `domain = kernel.space` if possible. However, if `kernel` is given as a function, there is no way to deduce a reasonable domain. 
Note also that the convolution of a function with support in `[a, b]` with a kernel supported in `[-c, c]` is supported in `[a-c, b+c]`, so for small kernels roughly in the same domain. That's not true if the kernel support is not zero-centered, say `[c, d]`. Then one gets `[a+c, b+d]` with midpoint `m = (a+b)/2 + (c+d)/2`, i.e. the "main mass" of the convolved function is in `[a, b] + (c + d)/2`, the old interval shifted by the midpoint of the kernel support.

To make a long story short: I don't think it's natural to assume that input function and kernel are supported in the same spatial domain. Therefore I would keep the requirement to give both.